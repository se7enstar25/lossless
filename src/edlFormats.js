import fastXmlParser from 'fast-xml-parser';
import i18n from 'i18next';

import csvParse from 'csv-parse';
import pify from 'pify';
import sortBy from 'lodash/sortBy';

import { formatDuration } from './util/duration';

const csvParseAsync = pify(csvParse);

export async function parseCsv(str) {
  const rows = await csvParseAsync(str, {});
  if (rows.length === 0) throw new Error(i18n.t('No rows found'));
  if (!rows.every(row => row.length === 3)) throw new Error(i18n.t('One or more rows does not have 3 columns'));

  const mapped = rows
    .map(([start, end, name]) => ({
      start: start === '' ? undefined : parseFloat(start, 10),
      end: end === '' ? undefined : parseFloat(end, 10),
      name,
    }));

  if (!mapped.every(({ start, end }) => (
    (start === undefined || !Number.isNaN(start))
    && (end === undefined || !Number.isNaN(end))
  ))) {
    console.log(mapped);
    throw new Error(i18n.t('Invalid start or end value. Must contain a number of seconds'));
  }

  return mapped;
}

export function parseCuesheet(cuesheet) {
  // There are 75 such frames per second of audio.
  // https://en.wikipedia.org/wiki/Cue_sheet_(computing)
  const fps = 75;

  const { tracks } = cuesheet.files[0];

  function parseTime(track) {
    const index = track.indexes[0];
    if (!index) return undefined;
    const { time } = index;
    if (!time) return undefined;

    return (time.min * 60) + time.sec + time.frame / fps;
  }

  return tracks.map((track, i) => {
    const nextTrack = tracks[i + 1];
    const end = nextTrack && parseTime(nextTrack);

    return { name: track.title, start: parseTime(track), end };
  });
}


export function parsePbf(text) {
  const chapters = text.split('\n').map((line) => {
    const match = line.match(/^[0-9]+=([0-9]+)\*([^*]+)*/);
    if (match) return { time: parseInt(match[1], 10) / 1000, name: match[2] };
    return undefined;
  }).filter((it) => it);

  const out = [];
  chapters.forEach((chapter, i) => {
    const nextChapter = chapters[i + 1];
    out.push({ start: chapter.time, end: nextChapter && nextChapter.time, name: chapter.name });
  });
  return out;
}

// https://developer.apple.com/library/archive/documentation/AppleApplications/Reference/FinalCutPro_XML/VersionsoftheInterchangeFormat/VersionsoftheInterchangeFormat.html
export function parseXmeml(xmlStr) {
  const xml = fastXmlParser.parse(xmlStr);
  // TODO maybe support media.audio also?
  return xml.xmeml.project.children.sequence.media.video.track.clipitem.map((item) => ({ start: item.start / item.rate.timebase, end: item.end / item.rate.timebase }));
}

export function parseYouTube(str) {
  const regex = /(?:([0-9]{2,}):)?([0-9]{1,2}):([0-9]{1,2})(?:\.([0-9]{3}))?[^\S\n]+([^\n]*)\n/g;

  const lines = [];

  function parseLine(match) {
    if (!match) return undefined;
    const [, hourStr, minStr, secStr, msStr, name] = match;
    const hour = hourStr != null ? parseInt(hourStr, 10) : 0;
    const min = parseInt(minStr, 10);
    const sec = parseInt(secStr, 10);
    const ms = msStr != null ? parseInt(msStr, 10) : 0;

    const time = (((hour * 60) + min) * 60 + sec) + ms / 1000;

    return { time, name };
  }

  let m;
  // eslint-disable-next-line no-cond-assign
  while ((m = regex.exec(`${str}\n`))) {
    lines.push(parseLine(m));
  }

  const linesSorted = sortBy(lines, (l) => l.time);

  const edl = linesSorted.map((line, i) => {
    const nextLine = linesSorted[i + 1];
    return { start: line.time, end: nextLine && nextLine.time, name: line.name };
  });

  return edl.filter((ed) => ed.start !== ed.end);
}

export function formatYouTube(segments) {
  return segments.map((segment) => {
    const timeStr = formatDuration({ seconds: segment.start, showMs: false, shorten: true });
    const namePart = segment.name ? ` ${segment.name}` : '';
    return `${timeStr}${namePart}`;
  }).join('\n');
}
