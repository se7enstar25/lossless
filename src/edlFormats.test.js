/* eslint-disable no-undef */
import { parseYouTube } from './edlFormats';

it('parseYoutube', () => {
  const str = `
Jump to chapters:
00:00 Test 1
00:01 Test 2
00:02 00:57 double
00:01:01 Test 3
01:01:01.012 Test 4
00:01:01.012 Test 5
   01:01.012 Test 6
  :01:02.012 Test 7
00:57:01.0123 Invalid 2
00:57:01. Invalid 3
01:15: Invalid 4
0132 Invalid 5
00:03
00:04     
00:05 
`;
  const edl = parseYouTube(str);
  expect(edl).toEqual([
    { start: 0, end: 1, name: 'Test 1' },
    { start: 1, end: 2, name: 'Test 2' },
    { start: 2, end: 4, name: '00:57 double' },
    { start: 4, end: 5, name: '' },
    { start: 5, end: 61, name: '' },
    { start: 61, end: 61.012, name: 'Test 3' },
    { start: 61.012, end: 62.012, name: 'Test 6' },
    { start: 62.012, end: 3661.012, name: 'Test 7' },
    { start: 3661.012, end: undefined, name: 'Test 4' },
  ]);
});

it('parseYouTube eol', () => {
  const str = ' 00:00 Test 1\n00:01 Test 2';
  const edl = parseYouTube(str);
  expect(edl).toEqual([
    { start: 0, end: 1, name: 'Test 1' },
    { start: 1, end: undefined, name: 'Test 2' },
  ]);
});
