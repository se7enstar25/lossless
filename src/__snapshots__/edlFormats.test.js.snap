// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses pbf 1`] = `
Array [
  Object {
    "end": 189.053,
    "name": "label 1",
    "start": 104.612,
  },
  Object {
    "end": 476.194,
    "name": "label 2",
    "start": 300.448,
  },
  Object {
    "end": 704.264,
    "name": "label 3",
    "start": 567.075,
  },
  Object {
    "end": 855.455,
    "name": "label 4",
    "start": 704.264,
  },
]
`;

exports[`parses pbf 2`] = `
Array [
  Object {
    "end": 189.053,
    "name": "label 1",
    "start": 104.612,
  },
  Object {
    "end": 300.448,
    "name": "label 2",
    "start": 189.053,
  },
  Object {
    "end": 476.194,
    "name": "label 3",
    "start": 300.448,
  },
  Object {
    "end": 567.075,
    "name": "label 4",
    "start": 476.194,
  },
  Object {
    "end": 704.264,
    "name": "label 5",
    "start": 567.075,
  },
  Object {
    "end": 855.455,
    "name": "label 6",
    "start": 704.264,
  },
]
`;

exports[`parses pbf 3`] = `
Array [
  Object {
    "end": 189.053,
    "name": "A",
    "start": 104.612,
  },
  Object {
    "end": 476.194,
    "name": "B",
    "start": 300.448,
  },
  Object {
    "end": 704.264,
    "name": "C",
    "start": 567.075,
  },
]
`;

exports[`parses pbf 4`] = `
Array [
  Object {
    "end": 189.053,
    "name": "A",
    "start": 104.612,
  },
  Object {
    "end": 476.194,
    "name": "B",
    "start": 300.448,
  },
]
`;

exports[`parses xmeml - with multiple tracks 1`] = `
Array [
  Object {
    "end": 11,
    "start": 0,
  },
  Object {
    "end": 4.08,
    "start": 0.96,
  },
  Object {
    "end": 3.88,
    "start": 1.12,
  },
  Object {
    "end": 4.32,
    "start": 0.84,
  },
  Object {
    "end": 3.36,
    "start": 0.36,
  },
  Object {
    "end": 6.24,
    "start": 0.36,
  },
  Object {
    "end": 4.52,
    "start": 0.24,
  },
  Object {
    "end": 2.48,
    "start": 0.48,
  },
  Object {
    "end": 2.36,
    "start": 0.48,
  },
  Object {
    "end": 1.36,
    "start": 0.36,
  },
  Object {
    "end": 1.36,
    "start": 0.36,
  },
  Object {
    "end": 1.04,
    "start": 0.36,
  },
]
`;

exports[`parses xmeml 1 1`] = `
Array [
  Object {
    "end": 8.56,
    "start": 0,
  },
  Object {
    "end": 13.12,
    "start": 8.56,
  },
  Object {
    "end": 19.12,
    "start": 13.12,
  },
  Object {
    "end": 27.96,
    "start": 19.12,
  },
  Object {
    "end": 39.08,
    "start": 27.96,
  },
  Object {
    "end": 49.36,
    "start": 39.08,
  },
]
`;

exports[`parses xmeml 2 1`] = `
Array [
  Object {
    "end": 2033.5666666666666,
    "start": 5.1,
  },
  Object {
    "end": 3839.2833333333333,
    "start": 3017.7,
  },
  Object {
    "end": 4686.366666666667,
    "start": 3862.9333333333334,
  },
  Object {
    "end": 6310.016666666666,
    "start": 5324.066666666667,
  },
  Object {
    "end": 6959.516666666666,
    "start": 6928.15,
  },
  Object {
    "end": 8081.116666666667,
    "start": 6985.9,
  },
  Object {
    "end": 9750.05,
    "start": 8728.066666666668,
  },
  Object {
    "end": 10545.116666666667,
    "start": 9793.316666666668,
  },
  Object {
    "end": 10864.883333333333,
    "start": 10740.633333333333,
  },
]
`;
